app/services/notify_service.rb

```diff
@@ -11,126 +11,175 @@ class NotifyService < BaseService
+    end
 
-  def optional_non_follower?
-    @recipient.user.settings['interactions.must_be_follower']  && !@notification.from_account.following?(@recipient)
-  end
+    def domain_blocking?
+      @recipient.domain_blocking?(@sender.domain) && !following_sender?
+    end
 
-  def optional_non_following?
-    @recipient.user.settings['interactions.must_be_following'] && !following_sender?
-  end
+    def conversation_muted?
+      @notification.target_status && @recipient.muting_conversation?(@notification.target_status.conversation)
+    end
 
-  def message?
-    @notification.type == :mention
+    def following_sender?
+      @recipient.following?(@sender)
+    end
   end
 
-  def direct_message?
-    message? && @notification.target_status.direct_visibility?
-  end
+  class FilterCondition
+    NEW_ACCOUNT_THRESHOLD = 30.days.freeze
 
-  # Returns true if the sender has been mentioned by the recipient up the thread
-  def response_to_recipient?
-    return false if @notification.target_status.in_reply_to_id.nil?
+    NEW_FOLLOWER_THRESHOLD = 3.days.freeze
 
-    # Using an SQL CTE to avoid unneeded back-and-forth with SQL server in case of long threads
-    !Status.count_by_sql([<<-SQL.squish, id: @notification.target_status.in_reply_to_id, recipient_id: @recipient.id, sender_id: @notification.from_account.id, depth_limit: 100]).zero?
-      WITH RECURSIVE ancestors(id, in_reply_to_id, mention_id, path, depth) AS (
-          SELECT s.id, s.in_reply_to_id, m.id, ARRAY[s.id], 0
-          FROM statuses s
-          LEFT JOIN mentions m ON m.silent = FALSE AND m.account_id = :sender_id AND m.status_id = s.id
-          WHERE s.id = :id
-        UNION ALL
-          SELECT s.id, s.in_reply_to_id, m.id, st.path || s.id, st.depth + 1
-          FROM ancestors st
-          JOIN statuses s ON s.id = st.in_reply_to_id
-          LEFT JOIN mentions m ON m.silent = FALSE AND m.account_id = :sender_id AND m.status_id = s.id
-          WHERE st.mention_id IS NULL AND NOT s.id = ANY(path) AND st.depth < :depth_limit
-      )
-      SELECT COUNT(*)
-      FROM ancestors st
-      JOIN statuses s ON s.id = st.id
-      WHERE st.mention_id IS NOT NULL AND s.visibility = 3
-    SQL
-  end
+    def initialize(notification)
+      @notification = notification
+      @recipient = notification.account
+      @sender = notification.from_account
+      @policy = NotificationPolicy.find_or_initialize_by(account: @recipient)
+    end
 
-  def from_staff?
-    @notification.from_account.local? && @notification.from_account.user.present? && @notification.from_account.user_role&.overrides?(@recipient.user_role)
-  end
+    def filter?
+      return false if override_for_sender?
 
-  def optional_non_following_and_direct?
-    direct_message? &&
-      @recipient.user.settings['interactions.must_be_following_dm'] &&
-      !following_sender? &&
-      !response_to_recipient?
-  end
+      from_limited? ||
+        (@policy.filter_not_following? && not_following?) ||
```

Comment:
For this line and the next four - maybe pull out the "does the policy filter this and is the condition true" logic to private methods, and then in this `filter?` method we'd have a slightly more readable chain of or'd conditions.

Example - make this one `filtered_by_not_following_policy?` and have that method definition be `@policy.filter_not_following? && not_following?`